eclipse-theia/theia
# Refactor Channel API to an object graph based approach

#### What it does

Refactors the message RPC API to use object-graph based Channels. This is achieved by making the binary
encoding of messages an implementation detail of the channel implementation. Instead of exposing a binary `writeBuffer` to write messages to, a `Channel` now has to implement a `send` method which is responsible for transporting the message/object over the wire. This way the API is easier to use for adopters. A `Channel` is now just a transport channel that can be used to send/receive any Javascript object. How the objects are encoded for efficient transport is up to the channel implementation.
The main advantages of this change are:
- The messaging API becomes generic. Adopters can easily implement communication protocols that dont't use binary encoding (an example for this is the Debug Protocol which used string-based encoding)
- Multiplexing or tunneling becomes trivial: you just wrap the object to send in another object containing routing information
- The main service channel is no longer limited to the default RPC protocol. Adopters can use the respective `ConnectionProvider` to receive a new (multiplexed) service channel and establish any custom protocol on top of it

The main code changes are the following:
- Refactor `Channel` API by
  - Replacing the `getWriteBuffer` method with a `send` method
  - Providing a reusable `AbstractChannel` implementation that setup all required event emiters and takes care of their disposal
 - Refactoring/Cleaning up the messsage buffer API (`WriteBuffer` and `ReadBuffer`) by
   - Combining the interface defintion and default implementations into a single file
   - Renaming the default implementations to `WriteBufferImpl`/`ReadBufferImpl` (previously `Uint8ArrayWriteBuffer`/`Uint8ArrayReadBuffer`)
   - Removing the `commit` method from the `WriteBuffer` API as it not longer serves a purpose
 - Introduce a new `MessageCodec` resp. `BinaryMessageCodec` API that combines the old `MessageEncoder` & `MessageDecoder` (and cleans up/removes the no longer needed RPC specific encoding methods)
 - Refactor the `DebugChannel` API to remove the overhead of having to encode each twice (first to string than to binary). We can now directly use a String channel instead.
 - Refactor/cleanup `ChannelMultiplexerAPI`. We can now simply wrap messages from the subchannels in another object of the form `{ id:string, message:any}`
 - Clean up file structure.
    - Remove `message-rpc` subfolder in `@theia/core/src/common` and move contained classes to `messaging` folder
    - Remove no longer needed classes from `message-rpc`
    - Clean up default exports (`index.ts) files to only export main API types
    
     
 Contributed on behalf of STMicroelectronics


#### How to test

Service communication should work as before. It probably makes sense to restest the bug fixes for bugs we encountered 
previously by:
- Testing that debugging works expected (https://github.com/eclipse-theia/theia/pull/11268)
    1. Open a .spec.ts file in the repo and try the `Run Mocha Tests` debug configuration.
    2. Debug view should work as expected. (Variables are shown, debug, debug trace, breakpoins are hit etc.
    3. Try a dynamic Node configuration like 'Run current file'. Should also work as expected.
 - Ensure that logging works as expected (https://github.com/eclipse-theia/theia/pull/11294)
    1. Add this log to DebugSessionManager set currentSession:
     ```bash
     console.log('SENTINEL FOR SETTING THE CURRENT SESSION', current);
    ```

    2. Start a debug session inside the application.
    3. Observe that the frontend (browser) logs log the message as expected.
    4. Observe that no error logged in the frontend console
    5. Observe that the backend logs do not show the message
    ( Same behavior as with the old RPC protocol)

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
